contracts/contracts/Umbra.sol

```diff
@@ -198,11 +209,82 @@ contract Umbra is Ownable {
     bytes32 _r,
     bytes32 _s
   ) external {
+    _validateWithdrawSignature(_stealthAddr, _acceptor, _tokenAddr, _sponsor, _sponsorFee, UmbraHookable(0), "", _v, _r, _s);
+    _withdrawTokenInternal(_stealthAddr, _acceptor, _tokenAddr, _sponsor, _sponsorFee, UmbraHookable(0), "");
+  }
+
+  /**
+   * @notice Withdraw an ERC20 token payment on behalf of a stealth address via signed authorization
+   * @param _stealthAddr The stealth address whose token balance will be withdrawn
+   * @param _acceptor Address where withdrawn funds should be sent
+   * @param _tokenAddr Address of the ERC20 token being withdrawn
+   * @param _sponsor Address which is compensated for submitting the withdrawal tx
+   * @param _sponsorFee Amount of the token to pay to the sponsor
+   * @param _hook Contract that will be called after the token withdrawal has completed
+   * @param _data Arbitrary data that will be passed to the post-withdraw hook contract
+   * @param _v ECDSA signature component: Parity of the `y` coordinate of point `R`
+   * @param _r ECDSA signature component: x-coordinate of `R`
+   * @param _s ECDSA signature component: `s` value of the signature
+   */
+  function withdrawTokenAndCallOnBehalf(
+    address _stealthAddr,
+    address _acceptor,
+    address _tokenAddr,
+    address _sponsor,
+    uint256 _sponsorFee,
+    UmbraHookable _hook,
+    bytes memory _data,
+    uint8 _v,
+    bytes32 _r,
+    bytes32 _s
+  ) external {
+    _validateWithdrawSignature(_stealthAddr, _acceptor, _tokenAddr, _sponsor, _sponsorFee, _hook, _data, _v, _r, _s);
+    _withdrawTokenInternal(_stealthAddr, _acceptor, _tokenAddr, _sponsor, _sponsorFee, _hook, _data);
+  }
+
+  /// @dev low level withdrawal function that should only be called after safety checks
+  function _withdrawTokenInternal(
+    address _stealthAddr,
+    address _acceptor,
+    address _tokenAddr,
+    address _sponsor,
+    uint256 _sponsorFee,
+    UmbraHookable _hook,
+    bytes memory _data
+  ) internal {
     uint256 _amount = tokenPayments[_stealthAddr][_tokenAddr];
 
-    // also protects from underflow
+    // also protects from underflow and re-entrance
     require(_amount > _sponsorFee, "Umbra: No balance to withdraw or fee exceeds balance");
 
+    uint256 _withdrawalAmount = _amount - _sponsorFee;
+    delete tokenPayments[_stealthAddr][_tokenAddr];
+    emit TokenWithdrawal(_stealthAddr, _acceptor, _withdrawalAmount, _tokenAddr);
+
+    SafeERC20.safeTransfer(IERC20(_tokenAddr), _acceptor, _withdrawalAmount);
```

Comment:
> If I want the hook to do something with my tokens, the hook address and acceptor address need to be the same

I don't think this is necessarily true (though it would probably be common). You could, for example, have a hook contract that is already pre-approved by the acceptor for a `transferFrom`, as one contrived example.

Reply:
Hm, yea good point. The main thought I had here was this part though:

> my first impression when reviewing this PR was the the hook would always be the same as the acceptor, if using a hook contract

But since that isn't true, still might be worth documenting somewhere, since the functionality wasn't clear until I reached this line. Guess this is really just a documentation nit 😛

Reply:
Hmm, I think I'm actually confused about what you're confused about 🙃 

Can you clarify what you mean here? You're not saying the hook *should* be the same as the acceptor, but you thought it had to be? What gave you that impression, if so, and what would you add to the docs to make it clearer?

Reply:
> You're not saying the hook should be the same as the acceptor, but you thought it had to be? 

Correct, that's why I thought when I started reviewing the PR, but I don't know why I thought that! I think it's fine to just move on though since from the interface makes it clear the acceptor is an explicit parameter